### PR TITLE
Review Changes for #555 #556 #594

### PR DESCRIPTION
**Pull request by nblogist**
## Description
Created following routes and Added optional stash parameter to:
1. own stake over time
2. total stake over time
3. other stake over time
4. nominators over time

## Motivation and Context
Currently errors are thrown for /getRewards when no records are found. This is not the proper way to handle that event. If there are no records found, then we should return an empty array and ensure the client can handle no results in a graceful way. #594 

own stake over time
total stake over time
other stake over time
nominators over time
These routes will grab directly from HistoricalValidatorStatistics table. #556 

Add an optional address parameter to /getRewards, /getOffences, /getSlashes, /getImOnline, /getX routes.
Many of these routes will use the new HistoricalValidatorStatistics table (look at this database model to see what data is on it).
Other routes such as rewards/slashes should grab from ChainEvents directly.  #555 

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [x] yes, but I did not run tests
- [ ] no

### CODE REVIEW

**Comment by drewstone** [comment]
At first glance, we'll definitely need unit tests for all these new routes with over 90% coverage :)! Take a look at sample unit tests to get an idea of how to write them. They're pretty straightforward.

**Comment by drewstone** [comment]
Can you add more comments please @nblogist, with so many different edge cases there's not enough comments on this PR

**Comment by nblogist** [comment]
@drewstone Currently if a startDate &  endDate isn't given, the route returns data for 1 year; If there is anything to be done please let me know.

File: server/router.ts [hunk 1]
```diff
@@ -215,10 +213,10 @@ function setupRouter(app, models, viewCountCache: ViewCountCache, identityFetchC
   router.get('/getInvites', passport.authenticate('jwt', { session: false }), getInvites.bind(this, models));
   router.get('/getOffences', getOffences.bind(this, models));
   router.get('/getRewards', getRewards.bind(this, models));
-  router.get('/getOwnStakeOverTime', getOwnStakeOverTime.bind(this, models));
-  router.get('/getTotalStakeOverTime', getTotalStakeOverTime.bind(this, models));
-  router.get('/getOtherStakeOverTime', getOtherStakeOverTime.bind(this, models));
-  router.get('/getNominatorsOverTime', getNominatorsOverTime.bind(this, models));
+  router.get('/getExposureOverTime', getOwnStakeOverTime.bind(this, models));
+  router.get('/getExposureOverTime', getOtherStakeOverTime.bind(this, models));
```
Comment (drewstone) [inline server/router.ts]: route names cannot be reused like this

File: server/router.ts [hunk 2]
```diff
@@ -215,10 +213,10 @@ function setupRouter(app, models, viewCountCache: ViewCountCache, identityFetchC
...
+  router.get('/getExposureOverTime', getTotalStakeOverTime.bind(this, models));
```
Comment (drewstone) [inline server/router.ts]: route names cannot be reused like this

File: server/router.ts [hunk 3]
```diff
@@ -215,10 +213,10 @@ function setupRouter(app, models, viewCountCache: ViewCountCache, identityFetchC
...
+  router.get('/getExposureOverTime', getNominatorsOverTime.bind(this, models));
```
Comment (drewstone) [inline server/router.ts]: route names cannot be reused like this

File: server/routes/getExposureOverTime.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    endDate = new Date();
+  }
+
+  const where: any = {
+    '$ChainEventType.chain$': chain,
+  };
+
+  // if stash is given
+  // eslint-disable-next-line no-unused-expressions
+  stash && (where['$HistoricalValidatorStatistic.stash$'] = stash);
```
Comment (drewstone) [inline server/routes/getExposureOverTime.ts:40]: please avoid this syntax for backend code, you should just do `if (stash) where['$...'] = stash`

File: server/routes/getExposureOverTime.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  if (startDate && endDate) {
+    where.created_at = {
+      [Op.between]: [startDate, endDate]
+    };
+  }
+
+  stakeOverTime = await models.HistoricalValidatorStatistic.findAll({
```
Comment (drewstone) [inline server/routes/getExposureOverTime.ts:48]: should this be returned?

File: server/routes/getExposureOverTime.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    order: [
+      ['created_at', 'ASC']
+    ],
+    attributes: ['stash', 'exposure', 'block'],
+    include: [ { model: models.ChainEventType } ]
+  });
+};
+
+const getTotalStakeOverTime = async (models, req: Request, res: Response, next: NextFunction) => {
+  getStakeOverTime(models, req, res, next);
```
Comment (drewstone) [inline server/routes/getExposureOverTime.ts:59]: no stakeOverTime return value, this is confusing.

File: server/routes/getExposureOverTime.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    } else {
+      validators[key] = [event_data.block_number, event_data.exposure.total];
+    }
+  });
+
+  return res.json({ status: 'Success', result: { validators } });
+};
+
+const getOwnStakeOverTime = async (models, req: Request, res: Response, next: NextFunction) => {
+  getStakeOverTime(models, req, res, next);
```
Comment (drewstone) [inline server/routes/getExposureOverTime.ts:80]: no return value here.

File: server/routes/getExposureOverTime.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    } else {
+      validators[key] = [event_data.block_number, event_data.exposure.own];
+    }
+  });
+
+  return res.json({ status: 'Success', result: { validators } });
+};
+
+const getOtherStakeOverTime = async (models, req: Request, res: Response, next: NextFunction) => {
+  getStakeOverTime(models, req, res, next);
```
Comment (drewstone) [inline server/routes/getExposureOverTime.ts:101]: no return value here

File: server/routes/getExposureOverTime.ts (hunk 4 above)
Comment (nblogist) [reply server/routes/getExposureOverTime.ts:40]: Thanks for letting me know; I'll take care next time.

File: server/routes/getExposureOverTime.ts (hunk 8 above)
Comment (nblogist) [reply server/routes/getExposureOverTime.ts:101]: I used a global variable, but yeah the return approach is preferable, I made the changes in next commit.

File: server/router.ts (hunk 3 above)
Comment (nblogist) [reply server/router.ts]: Yeah, saw my mistake here, fixed it in the last commit. Thanks for your time to review it!

**Comment by drewstone** [comment]
> @drewstone Currently if a startDate & endDate isn't given, the route returns data for 1 year; If there is anything to be done please let me know.

Please default to 30 days

**Comment by drewstone** [comment]
Also can you please point this PR to `staking-ui` or explain why you are pegged to a feature branch of a merge? Lets continue to merge things into `staking-ui` and separately we will merge `master` into that branch.

**Comment by nblogist** [comment]
@drewstone I unwittingly pushed the commits from _906f454fc7caa0ff96b5a5c58aefe51711da9125_ to _0b86ee647873656b535d8acfe4047a9b8a448113_  on [**_staking-ui_**](https://github.com/hicommonwealth/commonwealth/commits/staking-ui)
to be able to raise a PR, I made a new branch [**_ tmp_staking-ui_post_raymond_merger_** ](https://github.com/hicommonwealth/commonwealth/commits/tmp_staking-ui_post_raymond_merger) from _6c043bf6cf8e1c420d048b53f2876ef9a7b7e7e8_ (a commit before my direct commits) at staking-ui to give state of the branch before my unintended direct commits on it. 
In the same way I made a new branch from _906f454fc7caa0ff96b5a5c58aefe51711da9125_ through _0b86ee647873656b535d8acfe4047a9b8a448113_ and called it [**_tmp_furqan-555-556-594_** ](https://github.com/hicommonwealth/commonwealth/commits/tmp_furqan-555-556-594) and made any changes requested on it.

If you suggest I can make a new PR from [**_tmp_furqan-555-556-594_** ](https://github.com/hicommonwealth/commonwealth/commits/tmp_furqan-555-556-594) to [**_staking-ui_**](https://github.com/hicommonwealth/commonwealth/commits/staking-ui)

**Comment by nblogist** [comment]
> Only thing left in this PR is /getImOnline route, and adding an optional stash parameter to it.
> 
Done

**Comment by drewstone** [comment]
@nblogist the tests are failing still.

**Comment by drewstone** [comment]
Ok merging this in because the models don't exist in this branch so tests are failing :/
